src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+
+                        // Trigger signature help after starting the snippet session
+                        //
+                        // TODO: Figure out why ISignatureHelpBroker.TriggerSignatureHelp doesn't work but this does.
+                        // https://github.com/dotnet/roslyn/issues/50036
+                        var editorCommandHandlerServiceFactory = componentModel.DefaultExportProvider.GetExportedValue<IEditorCommandHandlerServiceFactory>();
+                        var editorCommandHandlerService = editorCommandHandlerServiceFactory.GetService(TextView, SubjectBuffer);
+                        editorCommandHandlerService.Execute((view, buffer) => new InvokeSignatureHelpCommandArgs(view, buffer), nextCommandHandler: null);
+
+                        return true;
+                    }
+                    else
+                    {
+                        _state.Clear(forceClearSymbolInformation: true);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Creates a snippet for providing arguments to a call.
+        /// </summary>
+        /// <param name="methodName">The name of the method as it should appear in code.</param>
+        /// <param name="includeMethod"><see langword="true"/> to include the method name and invocation parentheses in
+        /// the resulting snippet; otherwise, <see langword="false"/> if the method name and parentheses are assumed to
+        /// already exist and the template will only specify the argument placeholders.</param>
+        /// <param name="parameters">The parameters to the method. If the specific target of the invocation is not
+        /// known, an empty array may be passed to create a template with a placeholder where arguments will eventually
+        /// go.</param>
+        /// <returns></returns>
+        private static XDocument CreateSnippet(string methodName, bool includeMethod, ImmutableArray<IParameterSymbol> parameters, CancellationToken cancellationToken)
+        {
+            XNamespace snippetNamespace = "http://schemas.microsoft.com/VisualStudio/2005/CodeSnippet";
+
+            var template = new StringBuilder();
+
+            if (includeMethod)
+            {
+                template.Append(methodName).Append('(');
+            }
+
+            var declarations = new List<XElement>();
+            foreach (var parameter in parameters)
+            {
+                if (declarations.Any())
+                {
+                    template.Append(", ");
+                }
+
+                // Create a snippet field for the argument. The name of the field matches the parameter name, and the
+                // default value for the field is provided by a call to the internal ArgumentValue snippet function. The
+                // parameter to the snippet function is a serialized SymbolKey which can be mapped back to the
+                // IParameterSymbol.
```

Comment:
thanks for the explanation.  however, i'm still missing the "why" of this.  i.e. why is it important that we be able to map back to this particular parameter symbol once ths snippet is actually inserted?  (i'm literally just dumb.  i'm sure the answer is obvious. however, having not written or used this feature yet, it's not readily apparent during review, and would be appreciated) :)

Reply:
> once ths snippet is actually inserted

Snippets have placeholders, and each placeholder has an initial value. The initial value comes from one of two sources:

* A default value provided by the snippet itself
* A callback "snippet function" which the editor calls to obtain the default value

In this case, we're using the latter. The callback function is constrained to the signature defined by the platform, which in this case is always a single `string`. Since the callback needs a parameter symbol to evaluate the default value, it uses a `SymbolKey` to obtain the string that gets passed.

Reply:
oooh, that seems clever and interesting.

Reply:
> is there any reason we can't just directly generate the XML file here with the arguments already filled in?

We don't know the signature to use until Signature Help tells us.

Reply:
💭 We could move some code from the function back into the client, but it increases coupling and would be even more difficult to understand.

Reply:
➡️ This would amount to a substantial rewrite of the way this works, so I'll file a follow-up issue to investigate.

Reply:
> We don't know the signature to use until Signature Help tells us.

Exactly -- this method isn't running until you know the signature, but by this you can just directly put the values in.

> but it increases coupling and would be even more difficult to understand.

I'd say this is far more coupled and more difficult to understand, since there's an implicit handshake between this code and the function implementation.

Reply:
➡️ Leaving changes here for the follow-up issue.